### ISSUE TITLE
Really slow performance with 20 million nodes and 90 million relationships

### ISSUE THREAD

**Issue by zanek** [RedisLabsModules/RedisGraph#306]
I imported some data  with 20 million nodes and 90 million relationships

When I do simple queries like (within Redis-cli ) : 

 GRAPH.QUERY CONS "MATCH (x:customer {label: 'jane'})-[:KNOWS]->(d:customer  ) RETURN x,d LIMIT 18"

The queries return in 200 ms up to 3 seconds. I can actually bet faster results querying this data from SQLite !!

I've created indexes on all the node properties so I dont understand why this is so slow now.

When I had about 5 million nodes, I was getting most results less than 50 ms

Also, all this data comes from a 3 gigabyte dataset , but when I load it in to Redis, its using 49GB in memory. That is a huge expansion in memory for a small dataset 👎

**Comment by swilly22** [comment]
Would you mind sharing the output of:
GRAPH.EXPLAIN CONS "MATCH (x:customer {label: 'jane'})-[:KNOWS]->(d:customer ) RETURN x,d LIMIT 18"

You've mentioned that you've created an index for every node property, how many indices have you created in total?
Obviously indices have memory overhead, and so you might reduce your overall graph memory consumption if you delete unnecessary indices.

**Comment by zanek** [comment]
I get this when running explain 

2) 1) "Query internal execution time: 505.733000 milliseconds"
(0.62s)

"Produce Results\n    Conditional Traverse\n        Conditional Traverse\n            Index Scan\n"

**Comment by swilly22** [comment]
This shows that indeed an index is being utilised, but what amazes me is how long did it took to construct the execution plan (without actually running it) 0.62 seconds, which is bizarre!
Can you please share your machine spec, also I would appreciate if you can send me your CSV files so I can look into this. thank you.

**Comment by swilly22** [comment]
I've noticed the execution plan you've published:
`Produce Results\n Conditional Traverse\n Conditional Traverse\n Index Scan\n`
Doesn't match the query you've published
`MATCH (x:customer {label: 'jane'})-[:KNOWS]->(d:customer ) RETURN x,d LIMIT 18`

An execution plan for such query would include only a single `Conditional Traverse` please make sure you run:

`GRAPH.EXPLAIN CONS "MATCH (x:customer {label: 'jane'})-[:KNOWS]->(d:customer ) RETURN x,d LIMIT 18"`

**Comment by zanek** [comment]
You are correct about the query, Most single conditional queries are initially 50-100ms and then 1ms after I do the first query (I guess its caching the execution plan or something)

I didnt use a CSV file to import the data because the bulk loader kept crashing on me. I wrote a program using the Redis API to insert the data. I could send you the URL to dump.rdb file if you want. Please private message me or give me your email address for a link

On this particular machine, I'm using Ubuntu (via the Windows Subsystem for Linux). My machine specs are : 
	Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
        64GB Ram

**Comment by swilly22** [comment]
@zanek not sure how to private message you...
I would prefer receiving the CSV files inorder to see if there's an issue with the bulk-loader or an error in the way the CSV file(s) been constructed.

To conclude, your queries average latency are around 1ms? (after issuing the first query)

**Comment by zanek** [comment]
I still see slow searches with anything beyond simple conditionals. I understand you need to see my data to reproduce, but I can't just post my data on a Github ticket.
